File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/SimpleTaskExecutor.java
package com.example.alexeyglushkov.taskmanager.task;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Executor;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Created by alexeyglushkov on 08.02.15.
 */
public class SimpleTaskExecutor implements TaskExecutor {

    private static final ThreadFactory sThreadFactory = new ThreadFactory() {
        private final AtomicInteger mCount = new AtomicInteger(1);

        public Thread newThread(Runnable r) {
            return new Thread(r, "AsyncTask #" + mCount.getAndIncrement());
        }
    };

    public SimpleTaskExecutor() {
    }

    @Override
    public void executeTask(final Task task) {
        final Task.Callback callback = task.getTaskCallback();
        final Thread thread = sThreadFactory.newThread(new Runnable() {
            @Override
            public void run() {
                if (!Tasks.isTaskCompleted(task)) { // the task could be already handled by task manager
                    task.startTask(callback);
                }
            }
        });

        thread.start();
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/SimpleTaskExecutor.kt
package com.example.alexeyglushkov.taskmanager.task

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

/**
 * Created by alexeyglushkov on 08.02.15.
 */
open class SimpleTaskExecutor : TaskExecutor {
    private val viewModelJob = Job()
    private val taskScope = CoroutineScope(Dispatchers.IO + viewModelJob)

    override fun executeTask(task: Task) {
        taskScope.launch {
            start(task)
        }
    }

    suspend fun start(task: Task): Any? {
        return suspendCoroutine {
            task.taskCallback = object : Task.Callback {
                override fun onCompleted(cancelled: Boolean) {
                    val error = task.taskError
                    if (error != null) {
                        it.resumeWithException(error)
                    } else {
                        it.resume(task.taskResult)
                    }
                }
            }
            task.startTask()
        }
    }

    companion object {
        private val sThreadFactory = object : ThreadFactory {
            private val mCount = AtomicInteger(1)

            override fun newThread(r: Runnable): Thread {
                return Thread(r, "AsyncTask #" + mCount.getAndIncrement())
            }
        }
    }
}
